### DISCUSSION TITLE
Getting email adress of my user.

### DISCUSSION THREAD

**Discussion by arminvburren** [fief-dev/fief#13] [Q&A] [answered]
Hello,
I've tried the web app example in your doc. I get a token, scope, and id.
Is there a way as well to get at least the email adress used for authentication ? So I can show it on my website. Otherwise i'd need to redirect to the website and ask for like a username or the email again I guess... 
How are you guys in general for a web app ?

**Comment by frankie567** [comment, marked as answer]
Hi @arminvburren 👋 Welcome to Fief!

I guess you are talking about this part of [the example](https://docs.fief.dev/integrate/python/fastapi/#web-application-example):

```py
@app.get("/protected", name="protected")
async def protected(
    access_token_info: FiefAccessTokenInfo = Depends(auth.current_user()),
):
    return HTMLResponse(
        f"<h1>You are authenticated. Your user ID is {access_token_info['id']}</h1>"
    )
```

When you have a valid access token, you can make an API request to Fief to get fresh user information. Something like this:

```py
@app.get("/protected", name="protected")
async def protected(
    access_token_info: FiefAccessTokenInfo = Depends(auth.current_user()),
):
    userinfo = await fief.userinfo(access_token_info["access_token"])
    return HTMLResponse(
        f"<h1>You are authenticated. Your user email is {userinfo['email']}</h1>"
    )
```

However, I agree this is not very convenient and a additional helper could be added for this purpose, possibly with some cache to avoid an API request every time. I'll add this to the backlog!

**Reply by arminvburren** [reply]
Ok perfect. Well I guess i sort of need it once to create the profile in my db tho so it's fine to have a API call.

That said, how was i supposed to find this code ? Is there a deeper doc or do I have to go through fief code ? I don't have problem with this if it's expected but I quickly looked before asking and did not really find anything.

**Reply by frankie567** [reply]
Great!

Well, it's not provided directly like this, but there is a reference of the different methods available on the Fief client: https://docs.fief.dev/integrate/python/#fief-reference

Were you aware of this? I'm not sure it's clear that the `Python` header is clickable and actually has content inside.

**Reply by arminvburren** [reply]
I missed it indeed :|.

**Comment by frankie567** [comment]
Hello @arminvburren 👋 

Just to inform you that I've released a new version for the Python client, [v0.11.0](https://github.com/fief-dev/fief-python/releases/tag/v0.11.0) which introduces a built-in way to get and cache user information.

Basically, you'll be able to do something like this directly:

```py
@app.get("/protected", name="protected")
async def protected(
    user: FiefUserInfo = Depends(auth.current_user()),
):
    return HTMLResponse(
        f"<h1>You are authenticated. Your user email is {user['email']}</h1>"
    )
```

You can have a look at the [full example](https://docs.fief.dev/integrate/python/fastapi/#caching-user-information) to see how caching is implemented there.

**Reply by arminvburren** [reply]
Thank you, seems interesting and I may implement it later.

That said, I have been having a problem of slowness of my website recently and I started timing various parts. The main fief function which includes the call to fief.dev doesn't seem to be the biggest reason. Right now some of my pages take 1 to 2 seconds to load, even the most simple ones with only text. The Fief call is around 0.15s most of the time, even if I have peaks at 0.5 and even 1.5s some times right now.

Which lead me to a question: do you know how I could time and find what is slowing down my website ? I'm confused betweens calls to google cdn for stuff like bootstrap / ajax or the analytics, cloudflare (for the ssl), the database, and heroku. Is there a good tool for that ?

By the way the website is https://simu-immo.fr/ if you want to see your product in action :-).

**Reply by arminvburren** [reply]
Ok so already I found ways to improve speed time by removing extra js and doing more image compression thantks to https://pagespeed.web.dev/. If you have another input that would be cool but less critical.

So I've implemented the cache part but i'm not sure it's working in the intended way.

So far, in each of my functions has this shape:
```
@app.get("/simulation")
async def simulation_get(request: Request, user_session_cookie = Cookie(None)):
    auth_infos = await get_authentification_infos(request, user_session_cookie, db)
    template = blah()
    return template
```
They all call a get_authentification_infos(user_session_cookie) function which contains 
```

if user_session_cookie is not None:
       userinfo = await fief.userinfo(user_session_cookie).
```

Where exactly should I change with "auth.current_user()" which seem to be the new thing ? I don't really wanna use depends since my pages have to show even if I'm not connected.

**Reply by frankie567** [reply]
Hi @arminvburren 👋 

I think indeed making an HTTP request to Fief on each request will slow down your app quite a bit. You'll probably benefit from using the new `current_user` dependency with proper caching as described here: https://docs.fief.dev/integrate/python/fastapi/#caching-user-information

To help you with your use-case, I've just released a new version of the client, [`v0.14.0`](https://github.com/fief-dev/fief-python/releases/tag/v0.14.0), with support for the `optional` parameter. This way, you can define routes that retrieve the authenticated user if there is one, but still run if there is none: 

```py
@app.get("/protected", name="protected")
async def protected(
    user: Optional[FiefUserInfo] = Depends(auth.current_user(optional=True)),
):
    if user is None:
        return HTMLResponse(
            f"<h1>You are an anonymous user.</h1>"
        )
    return HTMLResponse(
        f"<h1>You are authenticated. Your user email is {user['email']}</h1>"
    )

```

I invite you to try those and let me know what you think!

**Reply by arminvburren** [reply]
Mm this is not working and I'm not sure why I did wrong. Basically user always ends up being "None" even after connection to fief.

Here's the code of the fief part of my website (with keys removed)


```
import uuid
from datetime import datetime
from typing import Dict, Optional

from fastapi import Depends, FastAPI, HTTPException, Query, Request, Response, status
from fastapi.responses import HTMLResponse, RedirectResponse
from fastapi.security import APIKeyCookie
from fief_client import FiefAccessTokenInfo, FiefAsync, FiefUserInfo
from fief_client.integrations.fastapi import FiefAuth

fief = FiefAsync(  # !
    "https://simu-immo.fief.dev",
    "########## KEY ##############################,
    "########## KEY ##############################",
)

def make_https_if_not_local(redirect_uri):
    if not '127.0.0.1:8000' in redirect_uri or 'localhost:8000' in redirect_uri:
        print(redirect_uri)
        return redirect_uri.replace("http://", "https://")
    print(redirect_uri)
    return redirect_uri

class CustomFiefAuth(FiefAuth):  # !
    client: FiefAsync

    async def get_unauthorized_response(self, request: Request, response: Response):
        redirect_uri =  make_https_if_not_local(request.url_for("auth_callback"))
        auth_url = await self.client.auth_url(redirect_uri, scope=["openid"])
        print(auth_url)
        raise HTTPException(
            status_code=status.HTTP_307_TEMPORARY_REDIRECT,
            headers={"Location": auth_url},
        )

# Cache 
class MemoryUserInfoCache:  # 
    def __init__(self) -> None:
        self.storage: Dict[uuid.UUID, FiefUserInfo] = {}  # 
    async def get(self, user_id: uuid.UUID) -> Optional[FiefUserInfo]:  # 
        return self.storage.get(user_id)
    async def set(self, user_id: uuid.UUID, userinfo: FiefUserInfo) -> None:  # 
        self.storage[user_id] = userinfo

memory_userinfo_cache = MemoryUserInfoCache()

async def get_memory_userinfo_cache() -> MemoryUserInfoCache:  # 
    return memory_userinfo_cache


scheme = APIKeyCookie(name="user_session_cookie", auto_error=False)
auth = CustomFiefAuth(fief, scheme, get_userinfo_cache=get_memory_userinfo_cache)


###################
### Auth Router ###
###################

from fastapi import APIRouter

authrouter = APIRouter()

@authrouter.get("/auth-callback", name="auth_callback")
async def auth_callback(request: Request, response: Response, code: str = Query(...), 
                        memory_userinfo_cache: MemoryUserInfoCache = Depends(get_memory_userinfo_cache)):
    redirect_uri = make_https_if_not_local(request.url_for("auth_callback"))
    tokens, userinfo = await fief.auth_callback(code, redirect_uri)

    response = RedirectResponse(make_https_if_not_local(request.url_for("login")))
    response.set_cookie(
        "user_session_cookie",
        tokens["access_token"],
        expires=int(datetime.now().timestamp() + tokens["expires_in"]),
        httponly=True,
        secure=False,  # ❌ Set this to `True` in production !
    )

    await memory_userinfo_cache.set(uuid.UUID(userinfo["sub"]), userinfo)

    return response

@authrouter.get("/login", name="login")
async def login(access_token_info: FiefAccessTokenInfo = Depends(auth.current_user())):
    response = HTMLResponse(f"<script>window.close();\
                              if(window.opener.location.pathname.includes('simulation')) \
                              {{window.opener.document.forms['simu_params_form'].submit();}} else \
                              {{window.opener.location.reload();}} \
                              </script>")
    return response

@authrouter.get("/logout", name="logout")
async def logout(request: Request):
    logout_redirect = "http://localhost:8000/" if request.client.host == "127.0.0.1" else "https://simu-immo.fr/"
    logout_url = await fief.logout_url(logout_redirect)
    response = RedirectResponse(logout_url)
    response.delete_cookie("user_session_cookie")
    return response

```

and calls are done in ways:
```
@app.get("/")
async def root(request: Request, user = Depends(auth.current_user(optional=True)), db = Depends(get_db)):

    auth_infos = await get_authentification_infos(request, user, db)

    return templates.TemplateResponse("index.html", {"request": request,
                                                     "which_background": "home3b.webp",
                                                     "auth_infos": auth_infos})
```
in which user is always None

**Reply by frankie567** [reply]
Is your cookie named `user_session_cookie` correctly set after `auth_callback`?

As a side note, you probably don't need `make_https_if_not_local` function. `request.url_for` is designed to generate URL with the correct scheme stated in the request. In some deployment environments however, you may need to set the [`FORWARDED_ALLOW_IPS` environment variable to `*`](https://www.uvicorn.org/settings/#http): it'll allow Uvicorn to trust the proxy header and thus generate HTTPS URL.

**Reply by arminvburren** [reply]
> Is your cookie named user_session_cookie correctly set after auth_callback?

How can i verify ? Everything work well with the version where i manually fetch the cookie and then call fief.userinfo, just not with cureent user.

Actually i tried cleaning all cookies, connecting with the new code and currentuser was None. Then i literally reverted back my code and was automatically connected. (Meaning the cookie was actually set up, just not recognized somehow by currentuser).

Not sure also where in the new code it is supposed to get the cookie value from what it puts in cache.

**Reply by frankie567** [reply]
You can check in the browsers's developer console. For Chrome, it's in the Application tab -> Cookies:

<img width="931" alt="Capture d’écran 2022-07-15 à 08 24 37" src="https://user-images.githubusercontent.com/1144727/179164057-f57aed0c-2ea8-43c9-a6af-4e5119347593.png">

Check it's in the list and it has the right domain.

**Reply by arminvburren** [reply]
Hey, quick update I had another try at this and the caching is working perfectly now :).
As a result the website is visibly faster; I think I gained between 0.15s and 0.4s speed per page by not calling fief everytime.

Thanks again !
